### PR TITLE
Jetpack Connect: Access state only through selectors

### PR DESCRIPTION
**Pull request by tyxla**
This PR aims to change the Jetpack Connect components to access the state only by using selectors. Currently, there are many places where state is being manipulated directly.

To achieve this, the PR introduces several more selectors and integrates them within the JPC flows.  Tests are also included.

To test:

* Checkout this branch.
* Verify that there are no regressions within any of the JPC flows.
* Verify that all JPC state tests pass: `npm run test-client client/state/jetpack-connect/`.

### CODE REVIEW

**Comment by ebinnion** [comment]
SSO code looks good and tests well to me. 👍

File: client/state/jetpack-connect/selectors.js [hunk 1]
```diff
@@ -1,16 +1,49 @@
...
+
+/**
  * Internal dependencies
  */
 import { getSiteByUrl } from 'state/sites/selectors';
 
 const JETPACK_CONNECT_TTL = 60 * 60 * 1000; // an hour
 
+const getConnectingSite = ( state ) => {
+	return get( state, [ 'jetpackConnect', 'jetpackConnectSite' ] );
```
Comment (johnHackworth) [inline client/state/jetpack-connect/selectors.js]: I may be wrong here, but I think we don't need to use `get`, since state.jetpackConnect is going to always be there

File: client/state/jetpack-connect/test/selectors.js [hunk 2]
```diff
@@ -6,49 +6,257 @@ import { expect } from 'chai';
...
+					alreadyConnected: false
+				}
+			};
+			const state = {
+				jetpackConnect: {
+					jetpackConnectSite
+				}
+			};
+
+			expect( getConnectingSite( state ) ).to.eql( jetpackConnectSite );
```
Comment (johnHackworth) [inline client/state/jetpack-connect/test/selectors.js]: now I see this test, I think it would be a good idea if this should be a failing one. 

I'm not sure if we already cover this in our standardized data approach ( @aduth ? ), but I think selectors should return copies of the state subtrees, not a pointer to the subtree. That way the components couldn't modify anything in the state. 

That way, this test should be failing ... what the selector returns shouldn't be the same object that is stored on the state, just another one with the same content

File: client/state/jetpack-connect/selectors.js (hunk 1 above)
Comment (tyxla) [inline client/state/jetpack-connect/selectors.js]: Well, `jetpackConnectSite` might be missing anyway, so I think for consistency we could just stick with `get()`.

File: client/state/jetpack-connect/test/selectors.js (hunk 2 above)
Comment (johnHackworth) [inline client/state/jetpack-connect/test/selectors.js]: oh, but if your selectors return a copy of the state, if you want to modify anything there directly from the code, you need to do it on purpose (passing the state, modifying it, not using redux anymore). The problem here is that if you return a pointer to the real state tree, we could modify the state by mistake, as a side effect. It's too easy to just assume that when you do 
` const site = getConnectingSite( state ) `

you are creating a local variable, so you can just do  `site.url = null` without any repercusion. And you would be changing the state directly

Comment (tyxla) [inline client/state/jetpack-connect/test/selectors.js]: I get your point and you're perfectly right, but there must be a reason why we're not doing that anywhere in our selectors. @aduth @mtias do you guys have any idea why we return a reference to the state in our selectors, and not a copy?

Anyway I think this is outside of the scope of this PR, as it would affect a lot of other selectors throughout Calypso as well.

File: client/signup/jetpack-connect/authorize-form.jsx [hunk 3]
```diff
@@ -636,34 +642,35 @@ const JetpackConnectAuthorizeForm = React.createClass( {
 
 export default connect(
 	state => {
-		const queryObjectSite = state.jetpackConnect.jetpackConnectAuthorize &&
-			state.jetpackConnect.jetpackConnectAuthorize.queryObject &&
-			state.jetpackConnect.jetpackConnectAuthorize.queryObject.site;
-
-		const site = queryObjectSite
-			? getSiteByUrl( state, state.jetpackConnect.jetpackConnectAuthorize.queryObject.site )
+		const authorizeSite = getAuthorizeSite( state );
```
Comment (tyxla) [inline client/signup/jetpack-connect/authorize-form.jsx]: Yeah, that's a good point, these 3 selectors (`getAuthorize*`) can be a little more informative.

File: client/signup/jetpack-connect/authorize-form.jsx [hunk 4]
```diff
@@ -636,34 +642,35 @@ const JetpackConnectAuthorizeForm = React.createClass( {
...
+		const site = authorizeSite
+			? getSiteByUrl( state, authorizeSite )
```
Comment (tyxla) [inline client/signup/jetpack-connect/authorize-form.jsx]: Good point 👍

**Comment by tyxla** [comment]
I've addressed the points that @johnHackworth mentioned, this is ready for a new review.

File: client/state/jetpack-connect/test/selectors.js (hunk 2 above)
Comment (johnHackworth) [inline client/state/jetpack-connect/test/selectors.js]: yeah, you're right, let's discuss this outside of this PR :+1:

**Comment by johnHackworth** [comment]
ok, ready to go... :ship: !

File: client/state/jetpack-connect/test/selectors.js (hunk 2 above)
Comment (ockham) [inline client/state/jetpack-connect/test/selectors.js]: `expect( something ).to.eql( somethingElse )` is deep comparison, while
`expect( something ).to.equall( somethingElse )` is flat (i.e. reference comparison for objects).

Yeah, bad choice of naming.

Comment (aduth) [inline client/state/jetpack-connect/test/selectors.js]: There's also `.to.deep.equal(` if you want to be more explicit.
